Suspend Parser for RenameRefactoring of Modules and Projects

BZngr opened the pull request:
This PR addresses the following FIXME comment found in `RenameRefactoring.cs`

> //FIXME: The parser needs to be suspended during the refactoring in case a
> component (or project) gets renamed because the VBE API object rename
> causes a separate reparse

.

bclothier commented on Rubberduck.Refactorings/Rename/RenameRefactoring.cs:
```diff
@@ -330,6 +334,16 @@ private void RenameVariable(RenameModel model, IRewriteSession rewriteSession)
+            var suspendResult = _parseManager.OnSuspendParser(this, new[] { ParserState.Ready }, () => RenameModuleInternal(model, rewriteSession));
+            if (suspendResult != SuspensionResult.Completed)
+            {
+                _logger.Warn("RenameModule failed.");
```
This is really a nitpick but I would prefer that the message was a bit more descriptive. I would like something more like:
```
_logger.Warn("RenameModule failed because a parser suspension request could not be fulfilled. The request's result was '{suspendResult.ToString()}'.");
```

bclothier commented on Rubberduck.Refactorings/Rename/RenameRefactoring.cs:
```diff
@@ -397,8 +411,18 @@ private void RenameModule(RenameModel model, IRewriteSession rewriteSession)
+            var suspendResult = _parseManager.OnSuspendParser(this, new[] { ParserState.Ready }, () => RenameProjectInternal(model, rewriteSession));
+            if (suspendResult != SuspensionResult.Completed)
+            {
+                _logger.Warn("RenameProject failed.");
```
Same comment as above.

BZngr commented:
I've modified the logger message per the suggestion above in addition to throwing an `OperationCanceledException` using the `RefactorFailure_BaseMessage` resource.  

I wrote the Suspend-Parser code for `RenameRefactoring` based on what I found in the `ExtractInterfaceRefactoring`.   The above comments would apply to the `ExtractInterfaceRefactoring` as well.  So, I've modified its Suspend-Parser code in the same manner.

MDoerner commented on Rubberduck.Refactorings/ExtractInterface/ExtractInterfaceRefactoring.cs:
```diff
@@ -67,7 +68,8 @@ private void AddInterface(ExtractInterfaceModel model)
             {
-                _logger.Warn("Extract interface failed.");
+                _logger.Warn($"Extract interface failed because a parser suspension request could not be fulfilled.The request's result was '{suspendResult.ToString()}'.");
+                throw new OperationCanceledException(RubberduckUI.RefactoringFailure_BaseMessage);
```
Please note that this will not notify the user. Instead it will be logged again as an error by `CommandBase`.

If you want to notify the user, you need to use a (possibly new) subclass of `RefactoringException`. Preferably, the corresponding notifier would be updated, too. (Messages presented to the user are not a concern of the refactoring itself.)

MDoerner commented on Rubberduck.Refactorings/Rename/RenameRefactoring.cs:
```diff
@@ -397,8 +413,19 @@ private void RenameModule(RenameModel model, IRewriteSession rewriteSession)
+            if (suspendResult != SuspensionResult.Completed)
+            {
+                _logger.Warn($"RenameProject failed because a parser suspension request could not be fulfilled.The request's result was '{suspendResult.ToString()}'.");
+                throw new OperationCanceledException(RubberduckUI.RefactoringFailure_BaseMessage);
```
Same as for the other exception.

MDoerner commented on Rubberduck.Refactorings/Rename/RenameRefactoring.cs:
```diff
@@ -330,6 +335,17 @@ private void RenameVariable(RenameModel model, IRewriteSession rewriteSession)
+            if (suspendResult != SuspensionResult.Completed)
+            {
+                _logger.Warn($"RenameModule failed because a parser suspension request could not be fulfilled.The request's result was '{suspendResult.ToString()}'.");
+                throw new OperationCanceledException(RubberduckUI.RefactoringFailure_BaseMessage);
```
Same as for the other exception.

BZngr commented:
I've added a `SuspendParserFailureException `based on prior comments.  A couple questions about `OnSuspendParser(...)`

>         /// <summary>
>         /// Suspends the parser for the action provided after the current parse has finished.
>         /// Any incoming parse requests will be executed afterwards.
>         /// </summary>
>         /// <param name="requestor">The object requesting a reparse.</param>
>         /// <param name="allowedRunStates">The states in which the action may be performed.</param>
>         /// <param name="busyAction">The action to perform.</param> 
>         /// <param name="millisecondsTimeout">The timeout for waiting on the current parse to finish.</param> 
> 

The millisecond timeout default of -1: I'm assuming this interpreted as 'infinite'(true?) - if so, I'm wondering if a limited timeout would be appropriate here.  Also, from the summary above, it sure looks like the additional parse we are trying to avoid - just gets queued and executed after the suspension.

bclothier commented:
@BZngr -- some background info:

The main reason it's set to `-1` is because when I wrote the suspension logic, we didn't know yet if there might be a case where suspension might cause the host to lock up. So far, it hasn't been reported, but since it was trivial to provide the timeout, I included it but used infinite timeout as the default since we anticipate that all suspensions would be short in duration (meaning milliseconds) and there should be safeguards in place to prevent it from locking itself up.

You are correct that suspension will simply queue any new requests which will be then processed after the suspension. The reason for this has to do with the fact that requests can come at any time, and do so asynchronously. Prior to the suspension logic being written, any new request would immediately cancel the current parse (if any) and start the parsing from start, to ensure that we got all the changes that needs to be considered for re-parsing. That was not changed when we introduced suspension since this is working well and handles the situation where a new parse request comes just right after the suspension ends, and behaves the same way as when it comes just right before the suspension ends. 

I hope that help clarifies the why.

BZngr commented:
@this Thank for the background.  My understanding of the purpose for the parser suspension was definitely incorrect.  So - I'm thinking the latest update addresses all the comments thus far.

bclothier commented on Rubberduck.Resources/RubberduckUI.resx:
```diff
@@ -1547,7 +1547,7 @@ NOTE: Restart is required for the setting to take effect.</value>
     <comment>{0}: name of module</comment>
   </data>
-  <data name="RenameFailure_TargetContriolNotFound" xml:space="preserve">
+  <data name="RenameFailure_TargetControlNotFound" xml:space="preserve">
```
I think you should update the bad key from other localized resource files or it'll break in non-English installs.

BZngr replied:
Thank you -> really good catch (I have to start thinking more globally when writing RD code).  Updated where the keys existed (missing in spanish and french)

bclothier commented on Rubberduck.Core/UI/Command/Refactorings/Notifiers/ExtractInterfaceFailedNotifier.cs:
```diff
@@ -22,6 +22,9 @@ protected override string Message(RefactoringException exception)
                         DeclarationType.ClassModule);
+                case SuspendParserFailureException suspendParserFailure:
+                    Logger.Warn(suspendParserFailure);
+                    return Resources.RubberduckUI.RefactoringFailure_SuspendParserFailure;
```
This is outside the PR's scope and would be another PR -- I want to note that I think that this should be in the base. However, the problem seems to be that the `Messge` gets overridden, rather than extended, so we are duplicating the common handling. In future, we should enhance so that `Message` can be extended, allowing the notifiers to focus on specialized concerns.

BZngr replied:
I've moved the handling of this exception case to `RefactoringFailureNotifierBase`.

MDoerner commented:
Quick question: have you tested renaming a module that contains a qualified reference to a member in the same module? I think introducing the suspension the way you did might change the outcome of the race condition still present in the code.

Probably, the VBE still lets us finish rewriting before it triggers the renamed event. If you really wanted to do the refactoring under parser suspension, you would have to include the actual rewrite in the suspension action, which suppressing the rewrite in the root method for this special case.

BZngr commented:
@MDoerner Thanks for spotting this.  I've reorganized the RenameRefactoring of Projects and Modules so that the rewrite is included in the suspension action.  This PR updates both `ExtractInterface` and `Rename` - the PR's `ExtractInterface` changes were already performing the rewrite in the suspension action.  

I find myself wondering if there is any downside (and possibly an advantage) to suspending the parser during the execution of all refactorings.  If so, then the suspension could possibly be built into the base class as part of calling to `RefactorImpl`.

bclothier commented:
Regarding the question - yes all refactoring should be in a suspension action since by definition, it will be altering some kind of code module or object, which makes it unsafe to do in a middle of a parse. That said, this can be done in a new PR - the rename refactoring is used quite commonly so it'll be a good test for how well suspension works out in real world before we make all refactoring adhere to it.

MDoerner commented on Rubberduck.Refactorings/Rename/RenameRefactoring.cs:
```diff
@@ -13,6 +13,8 @@
 using Rubberduck.VBEditor.SafeComWrappers;
 using Rubberduck.VBEditor.Utility;
+using NLog;
+using Rubberduck.Resources;
```
Is this used anywhere?

BZngr replied:
`using NLog `is there due to a logged warning message when there is a parser suspension failure.  The logged warning was present in parser suspension code of `ExtractInterface` which I used as the model for the `RenameRefactoring` changes.  So - `using NLog` is occurs in both.

`using Rubberduck.Resources` is not used (I'll clean that up on my next sync).

BZngr replied:
Regarding earlier comment 

> the rename refactoring is used quite commonly so it'll be a good test for how well suspension works out in real world before we make all refactoring adhere to it

The PR suspends the parser for renaming Modules and Projects only - which I'm suspecting is used less frequently than other renaming actions.  Is there interest in modifying the logic to suspend the parser for all rename refactoring types?

MDoerner replied:
I have to say that I do not agree that refactorings should always work under parser suspension. For this PR, having it only for the module and project renames is fine.

retailcoder replied:
The idea is to prevent VBIDE events from triggering a reparse before the refactoring is ready to trigger it; renaming a project or module normally triggers a reparse through custom events, so suspending shuts down that response and lets the refactoring do its job (alternatively we could stop firing these events). This isn't needed in other rename scenarios.

MDoerner replied:
I did not doubt that it makes sense for these scenarios. My doubts is about generally suspending in refactorings, even if the refactoring does not cause a separatebreparse itself.

bclothier replied:
The thing is that by definition, refactoring means rewriting, which means a change in the code module and therefore triggers a reparse?

I am having trouble imagining what a refactoring that didn't require parse would look like.

MDoerner replied:
If the refactoring does not cause separate reparses via VBE events, we already guarantee that the state is consistent via the rewrite session.

BZngr replied:
So, if I was to summarize the above: Any refactoring that renames or creates a module or project should suspend the parser during the refactoring.  So, only the `Rename `refactoring and `ExtractInterface `refactoring meet this criteria and contain parser suspension code.
